src/v/kafka/server/group_router.h

```diff
@@ -69,6 +69,33 @@ class group_router final {
           });
     }
 
+    template<typename Request, typename FwdFunc>
+    auto route_tx(Request&& r, FwdFunc func) {
+        // get response type from FwdFunc it has return future<response>.
+        using return_type = std::invoke_result_t<
+          FwdFunc,
+          decltype(std::declval<group_manager>()),
+          Request&&>;
+        using resp_type = typename return_type::value_type;
+
+        auto m = shard_for(r.group_id);
+        if (!m) {
+            resp_type reply;
+            reply.ec = cluster::tx_errc::not_coordinator;
+            return ss::make_ready_future<resp_type>(reply);
+        }
+        r.ntp = std::move(m->first);
+        return with_scheduling_group(
+          _sg, [this, func, shard = m->second, r = std::move(r)]() mutable {
+              return _group_manager.invoke_on(
+                shard,
+                _ssg,
+                [func, r = std::move(r)](group_manager& mgr) mutable {
+                    return std::invoke(func, mgr, std::move(r));
```

Comment:
should we assert that `r` is in a foreign_ptr<> otherwise, isn't this violating the x-core mechanics here ? i.e.: the owning core, will not get a chance to get notified about the reclamation.

Reply:
We already have exactly the same situation with `route`, I've just adjusted it for a new data type and called `route_tx`. Do you think I missed something or is it a problem for both templates?

Reply:
technically, it works but it's relatively expensive on the seastar memory reclamation. Usually the service and templates should do something like this

https://github.com/vectorizedio/redpanda/blob/082a93c03de23a52326e28b7f51ac413b854cb75/src/v/raft/service.h#L139


From a mechanical perspective one usually does a  `unique_ptr<foreign_ptr<Request>>` you can do a type assertion if you want. check out the record_batch_reader::data_t or other examples. happy to brainstorm too.

Reply:
Thanks, I understand it now. But still we already have this issue with `route` and fixing it now only make the scope of this PR larger, created an [issue](https://github.com/vectorizedio/redpanda/issues/1204) to fix it later.